ibc/docs/ics23.md

```diff
@@ -0,0 +1,78 @@
+# ICS-23 Vector Commitments <!-- omit in toc -->
+
+- [Overview](#overview)
+- [Implementation](#implementation)
+  - [Custom SMT `ProofSpec`](#custom-smt-proofspec)
+  - [Converting `SparseMerkleProof` to `CommitmentProof`](#converting-sparsemerkleproof-to-commitmentproof)
+  - [Proof Verification](#proof-verification)
+
+## Overview
+
+[ICS-23][ics23] defines the types and functions needed to verify membership of a key-value pair in a `CommitmentState`. As the Pocket IBC implementation uses the [SMT][smt] for its provable stores this is what is referred to as the `CommitmentState` object. Cosmos has a library `cosmos/ics23` which is already SDK agnostic and defines many of the types necessary for ICS-23. This library was able to be used _mostly_ out of the box, with some minor adjustments detailed below.
+
+## Implementation
+
+The benefit of using `cosmos/ics23` over implementing similar types ourselves is twofold:
+
+1. It is already SDK agnostic, so can be used by Pocket (a non-cosmos chain) without any issues.
+2. The functions defined for proof verification are separate from the tree structures themselves
+   - This means we do not need to interact with an SMT instance in order to verify proofs.
+
+However, there were some changes made specifically for Pocket's implementation of ICS-23.
+
+### Custom SMT `ProofSpec`
+
+The `ProofSpec` type in `cosmos/ics23` is used to define the steps needed to verify a proof, what hashing functions should be used, any node prefixes, etc. This is then passed into the verification functions in order to verify a proof, instead of having to interact with the tree itself. This is useful as proofs must be verified on a light-client and as such being able to verify a proof without creating a tree is much more memory efficient.
```

Comment:
No so remember the way that we construct a proof requires us to provide the hashes of the neighbouring nodes so that when we hash the key-value pair in conjunction with the hashes in the proof (as long as we get their left/right order correct) we can reconstruct the root hash. The tree itself does this method without going back "into" the tree. What the using the tree does is that it allows us to use the hashing functions the tree was set up with without providing them again.

So when we do this without the tree it means
1. We do not need to create a tree
2. We follow the exact same steps
3. We have to provide the hashing functions
4. In the "steps" we have to say the order of the hashes

If we look at how we define the order of the hashes its kinda simple:
```go
func convertSideNodesToSteps(sideNodes [][]byte, path []byte) []*ics23.InnerOp {
	steps := make([]*ics23.InnerOp, 0, len(sideNodes))
	for i := 0; i < len(sideNodes); i++ {
		var prefix, suffix []byte
		prefix = append(prefix, innerPrefix...)
		if getPathBit(path, len(sideNodes)-1-i) == left {
			suffix = make([]byte, 0, len(sideNodes[i]))
			suffix = append(suffix, sideNodes[i]...)
		} else {
			prefix = append(prefix, sideNodes[i]...)
		}
		op := &ics23.InnerOp{
			Hash:   ics23.HashOp_SHA256,
			Prefix: prefix,
			Suffix: suffix,
		}
		steps = append(steps, op)
	}
	return steps
```

Then when we hash using this step with either a prefix or a suffix we add the current hash into the missing space and we have an inner node! This is then hashed to produce the next current hash and so on until we have the root.

Reply:
Got it. So the tl;dr of the point you're making is that merkle proofs are verifiable without the tree. Correct?